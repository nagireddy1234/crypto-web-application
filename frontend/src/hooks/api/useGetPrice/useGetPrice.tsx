import { settings } from "@/settings";
import { PriceData } from "@/types/price";
import { useQuery } from "@tanstack/react-query";

const getPrice = async (currency: string): Promise<PriceData> => {
  const response = await fetch(
    `${settings.api.baseUrl}/price?symbol=${currency}`
  );

  if (!response.ok) {
    throw new Error(
      `Error fetching price for ${currency}, status: ${response.status}`
    );
  }

  const data: PriceData = await response.json();
  return data;
};

export const useGetPrice = (currency: string) => {
  const queryCacheTime =
    parseInt(import.meta.env.VITE_QUERY_CACHE_TIME) || 1800000; // Cache time (30 minutes)

  return useQuery({
    queryKey: ["price", currency],
    queryFn: () => getPrice(currency),
    staleTime: queryCacheTime,
    throwOnError: false,
    retry: false,
  });
};
