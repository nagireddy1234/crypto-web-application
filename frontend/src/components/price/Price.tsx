import { useState } from "react";
import { useGetPrice } from "@/hooks/api/useGetPrice";
import { Link } from "react-router-dom";
import { PriceCard } from "./PriceCard/PriceCard";

export const Price = () => {
  const [currency, setCurrency] = useState("TON/USDT");
  const { data, isError, isLoading } = useGetPrice(currency);

  return (
    <>
      <div className="container flex justify-between items-center mx-auto px-6 pt-10 pb-14 max-w-7xl">
        <div>
          <div className="flex items-center gap-3 mb-6">
            <img
              src="crypto.svg"
              alt="Logo"
              className="h-6 w-6 rounded-full shadow-lg"
              style={{
                transform: "rotate(45deg)",
                boxShadow:
                  "0 4px 8px rgba(0, 255, 0, 0.3), 0 0 15px rgba(0, 255, 0, 0.5)",
              }}
            />
            <h2 className="text-white text-lg font-bold">Crypto Converter</h2>
          </div>
        </div>
        <Link
          to="/history"
          className="text-blue-500 hover:underline hover:text-blue-400 transition duration-200 ease-in-out"
        >
          History
        </Link>
      </div>

      <div className="flex justify-center mb-8 space-x-6">
        {["TON/USDT", "ETH/BTC", "BTC/USDT"].map((pair) => (
          <label
            key={pair}
            className={`cursor-pointer text-white text-lg p-2 rounded-md transition-colors duration-200 ${
              currency === pair
                ? "bg-blue-500"
                : "bg-gray-700 hover:bg-gray-600"
            }`}
          >
            <input
              type="radio"
              name="currency"
              value={pair}
              checked={currency === pair}
              onChange={() => setCurrency(pair)}
              className="hidden"
            />
            <span>{pair}</span>
          </label>
        ))}
      </div>

      <div className="grid md:grid-cols-2 gap-10 max-w-7xl p-4 mx-auto my-12">
        {isError && (
          <div className="h-28 flex justify-center items-center text-red-500">
            <p>Error fetching data.</p>
          </div>
        )}
        {isLoading && (
          <div className="h-28 flex justify-center items-center text-gray-500">
            <p>Loading...</p>
          </div>
        )}
        {data &&
          Object.entries(data).map(([pair, price]) => (
            <PriceCard key={pair} pair={pair} price={price} />
          ))}
      </div>
    </>
  );
};
