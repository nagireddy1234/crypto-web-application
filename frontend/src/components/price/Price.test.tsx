import { render, screen, fireEvent } from "@testing-library/react";
import { Price } from "./Price";
import { useGetPrice } from "@/hooks/api/useGetPrice";
import { vi } from "vitest";
import { BrowserRouter } from "react-router-dom";

vi.mock("../../hooks/api/useGetPrice", () => ({
  useGetPrice: vi.fn(),
}));

describe("Price Component", () => {
  it("displays loading message when data is being fetched", async () => {
    (useGetPrice as jest.Mock).mockReturnValue({
      data: null,
      isLoading: true,
      isError: false,
    });

    render(
      <BrowserRouter>
        <Price />
      </BrowserRouter>
    );

    expect(screen.getByText("Loading...")).toBeInTheDocument();
  });

  it("displays error message when there is an error fetching data", async () => {
    (useGetPrice as jest.Mock).mockReturnValue({
      data: null,
      isLoading: false,
      isError: true,
    });

    render(
      <BrowserRouter>
        <Price />
      </BrowserRouter>
    );

    expect(screen.getByText("Error fetching data.")).toBeInTheDocument();
  });

  it("displays price cards when data is successfully fetched", async () => {
    const mockData = {
      TONUSDT: 1.23,
      USDTTON: 0.81,
    };
    (useGetPrice as jest.Mock).mockReturnValue({
      data: mockData,
      isLoading: false,
      isError: false,
    });

    render(
      <BrowserRouter>
        <Price />
      </BrowserRouter>
    );

    expect(screen.getAllByText("TONUSDT")[0]).toBeInTheDocument();
    expect(screen.getByText("1.230000")).toBeInTheDocument();
    expect(screen.getByText("USDTTON")).toBeInTheDocument();
    expect(screen.getByText("0.810000")).toBeInTheDocument();
  });

  it("changes the currency when currency is updated", async () => {
    const mockData = {
      ETHBTC: 1,
      BTCETH: 0.81,
    };
    (useGetPrice as jest.Mock).mockReturnValue({
      data: mockData,
      isLoading: false,
      isError: false,
    });

    render(
      <BrowserRouter>
        <Price />
      </BrowserRouter>
    );
    fireEvent.change(screen.getByText("ETH/BTC"));
    expect(screen.getByText("1")).toBeInTheDocument();
    expect(screen.getByText("BTCETH")).toBeInTheDocument();
    expect(screen.getByText("0.810000")).toBeInTheDocument();
    expect(screen.getAllByText("ETHBTC")[0]).toBeInTheDocument();
  });

  it("Snapshot test", () => {
    const mockData = {
      TONUSDT: 1.23,
      USDTTON: 0.81,
    };
    (useGetPrice as jest.Mock).mockReturnValue({
      data: mockData,
      isLoading: false,
      isError: false,
    });

    const { container } = render(
      <BrowserRouter>
        <Price />
      </BrowserRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
