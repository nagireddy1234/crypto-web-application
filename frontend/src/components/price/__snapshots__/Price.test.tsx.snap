// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Price Component > Snapshot test 1`] = `
<div>
  <div
    class="container flex justify-between items-center mx-auto px-6 pt-10 pb-14 max-w-7xl"
  >
    <div>
      <div
        class="flex items-center gap-3 mb-6"
      >
        <img
          alt="Logo"
          class="h-6 w-6 rounded-full shadow-lg"
          src="crypto.svg"
          style="transform: rotate(45deg); box-shadow: 0 4px 8px rgba(0, 255, 0, 0.3), 0 0 15px rgba(0, 255, 0, 0.5);"
        />
        <h2
          class="text-white text-lg font-bold"
        >
          Crypto Converter
        </h2>
      </div>
    </div>
    <a
      class="text-blue-500 hover:underline hover:text-blue-400 transition duration-200 ease-in-out"
      href="/history"
    >
      History
    </a>
  </div>
  <div
    class="flex justify-center mb-8 space-x-6"
  >
    <label
      class="cursor-pointer text-white text-lg p-2 rounded-md transition-colors duration-200 bg-blue-500"
    >
      <input
        checked=""
        class="hidden"
        name="currency"
        type="radio"
        value="TON/USDT"
      />
      <span>
        TON/USDT
      </span>
    </label>
    <label
      class="cursor-pointer text-white text-lg p-2 rounded-md transition-colors duration-200 bg-gray-700 hover:bg-gray-600"
    >
      <input
        class="hidden"
        name="currency"
        type="radio"
        value="ETH/BTC"
      />
      <span>
        ETH/BTC
      </span>
    </label>
    <label
      class="cursor-pointer text-white text-lg p-2 rounded-md transition-colors duration-200 bg-gray-700 hover:bg-gray-600"
    >
      <input
        class="hidden"
        name="currency"
        type="radio"
        value="BTC/USDT"
      />
      <span>
        BTC/USDT
      </span>
    </label>
  </div>
  <div
    class="grid md:grid-cols-2 gap-10 max-w-7xl p-4 mx-auto my-12"
  >
    <div
      class="bg-gradient-to-br from-blue-600 to-green-500 p-6 rounded-lg shadow-xl transition-transform duration-300 ease-out transform hover:scale-105 hover:shadow-2xl"
    >
      <h2
        class="text-lg font-semibold text-white tracking-wide mb-2"
      >
        TONUSDT
      </h2>
      <p
        class="text-4xl font-extrabold text-white tracking-tight"
      >
        1.230000
      </p>
    </div>
    <div
      class="bg-gradient-to-br from-blue-600 to-green-500 p-6 rounded-lg shadow-xl transition-transform duration-300 ease-out transform hover:scale-105 hover:shadow-2xl"
    >
      <h2
        class="text-lg font-semibold text-white tracking-wide mb-2"
      >
        USDTTON
      </h2>
      <p
        class="text-4xl font-extrabold text-white tracking-tight"
      >
        0.810000
      </p>
    </div>
  </div>
</div>
`;
